File: 7_kyu/Which_color_is_the_brightest.go
package kata

// Gets value of the brightes channel of the color.
// Color must be in format "#rrggbb" with hex values 00..FF
func brightestChannel(color string) string {
	if len(color) != 7 || color[0] != '#' {
		// Error exit
		return "#00"
	}
	accu := color[1:3]
	for n := 3; n < 6; n += 2 {
		t := color[n : n+2]
		if t > accu {
			accu = t
		}
	}
	return accu
}

// Returns color with brightes color channel.
// Returns first of maximal color channel is more than one in the list.
func Brightest(colors []string) string {
	if len(colors) < 1 {
		// Error exit
		return "#000000"
	}
	accu := colors[0]
	maxChannel := brightestChannel(accu)
	for i := 1; i < len(colors); i++ {
		m := brightestChannel(colors[i])
		if m > maxChannel {
			accu = colors[i]
			maxChannel = m
		}
	}
	return accu
}

File: 7_kyu/Which_color_is_the_brightest.rs
// Returns RGB string with maximal value.
// Returns first string if maximal value is in more than one string.
// Returns black if array is empty.
fn brightest(colors: &[String]) -> String {
    match colors.iter().rev().max_by_key(|&s| value(s)) {
        Some(m) => m.clone(),
        None => String::from("#000000"),
    }
}

// Returns value of a hex RGB string (#RRGGBB).
// Uses 0 as error fallback if some color code is invalid.
fn value(rgb: &String) -> u8 {
    if rgb.len() != 7 {
        return 0;
    };
    let r = u8::from_str_radix(&rgb[1..3], 16).unwrap_or(0);
    let g = u8::from_str_radix(&rgb[3..5], 16).unwrap_or(0);
    let b = u8::from_str_radix(&rgb[5..7], 16).unwrap_or(0);
    r.max(g).max(b)
}
